include/picongpu/traits/GetMargin.hpp

```diff
@@ -25,34 +25,56 @@ namespace picongpu
 {
     namespace traits
     {
-        /**Get margin of a solver
-         * class must define a LowerMargin and UpperMargin for any valid solver
-         *
-         * \tparam Solver solver which needs ghost cells for solving a problem
-         *         if solver not define `LowerMargin` and `UpperMargin` this trait (GetMargin)
-         *         must be specialized
-         * \tparam T_Parameter an optional parameter type
-         * for different objects
+        /** Trait to get margins of a (solver) type
+         *
+         * Defines internal types LowerMargin and UpperMargin.
+         * Both are compile-time simDim-dimentional integer vectors with the number of margin cells required.
+         * Thus, T_Type applied to any cell gives a guarantee to use not more than LowerMargin cells to the "left" and
+         * not more than UpperMargin cells to the "right" from the cell.
+         *
+         * By default propagates eponimous internal types of T_Type.
+         *
+         * @tparam T_Type solver type
+         * @tparam T_Parameter an optional parameter type
          */
-        template<class Solver, typename T_Parameter = void>
+        template<typename T_Type, typename T_Parameter = void>
         struct GetMargin
         {
-            using LowerMargin = typename Solver::LowerMargin;
-            using UpperMargin = typename Solver::UpperMargin;
+            using LowerMargin = typename T_Type::LowerMargin;
+            using UpperMargin = typename T_Type::UpperMargin;
         };
 
-        template<typename T_Type>
+        /** Trait to get a lower margin of a (solver) type
+         *
+         * Defines internal type which is a compile-time simDim-dimentional integer vector.
+         * T_Type applied to any cell gives a guarantee to use not more than type cells to the "left" from the cell.
+         *
+         * By default propagates GetMargin<...>::LowerMargin.
+         *
+         * @tparam T_Type solver type
+         * @tparam T_Parameter an optional parameter type
+         */
+        template<typename T_Type, typename T_Parameter = void>
         struct GetLowerMargin
         {
-            typedef typename traits::GetMargin<T_Type>::LowerMargin type;
+            using type = typename GetMargin<T_Type, T_Parameter>::LowerMargin;
         };
 
-        template<typename T_Type>
+        /** Trait to get an upper margin of a (solver) type
+         *
+         * Defines internal type which is a compile-time simDim-dimentional integer vector.
+         * T_Type applied to any cell gives a guarantee to use not more than type cells to the "right" from the cell.
+         *
+         * By default propagates GetMargin<...>::UpperMargin.
+         *
+         * @tparam T_Type solver type
```

Comment:
I suggest renaming the template parameter `T_Type` into `T_Solver` to be more specific.

Reply:
Done